kong/plugins/ddtrace/propagation.lua

```diff
@@ -1,68 +1,21 @@
--- Propagation methods for Datadog APM: extration and injection of datadog-specific request headers.
---
-local ffi = require "ffi"
-ffi.cdef[[
-unsigned long long int strtoull(const char *nptr, char **endptr, int base);
-]]
+local datadog = require "kong.plugins.ddtrace.datadog_propagation"
+local new_span = require "kong.plugins.ddtrace.span".new
 
-
-local function parse_uint64(str)
-    if not str then
-        return nil, "unable to parse, value is nil"
-    end
-    ffi.errno(0)
-    local parsed_str = ffi.C.strtoull(str, nil, 10)
-    local err = ffi.errno()
-    if err ~= 0 then
-        return nil, "unable to parse '" .. str .. "' as 64-bit number, errno=" .. err
-    end
-    -- TODO: check the entire string was consumed, instead of partially decoded
-    return parsed_str, nil
-end
-
-local function extract(headers)
-    local trace_id_value = headers["x-datadog-trace-id"]
-    if not trace_id_value then
-        -- no trace ID, therefore no propagated trace
-        return nil, nil, nil, nil, nil
+local function extract_or_create_span(request, span_options, max_header_size)
+    local trace_id, parent_id, sampling_priority, origin, tags, err = datadog.extract(request.get_header, max_header_size)
+    local span = new_span(span_options.service, span_options.name, span_options.ressource, trace_id, nil, parent_id, span_options.start_us, sampling_priority, origin, nil)
+    if tags then
```

Comment:
Might be good to choose a name that helps the reader distinguish between "tags" in the sense of X-Datadog-Tags versus "tags" in the sense of DD_TAGS.